### ISSUE TITLE
Opening a page with additional auth header overwrites auth header for XHR requests

### ISSUE THREAD

**Issue by F7502** [getgauge/taiko#816]
### Expected behavior

When I open a page and provide additional HTTP headers for authentication, all subsequent requests contain these auth headers, except XHR requests.

### Actual behavior

When I open a page and provide additional HTTP headers for authentication, this also interferes with XHR requests sent from the page to backend APIs. I do:

```javascript
goto('https://example.com',{headers:{'Authorization': 'Basic xxx'},waitForNavigation:false,navigationTimeout:10000})
click("Here")
click("There")
```
All following requests like e.g. to `https://example.com/my.css` have the same auth header set, which is fine. But, the page contains some javascript code that does XHR requests to a backend API, e.g. to `https://api.example.com`. These requests also have the basic auth header set, although the javascript code sets some different auth header (JWT as Bearer token), which apparently gets overwritten by taiko. Consequently, the backend calls do not work anymore.

(The page is protected with basic auth. To avoid the login popup, which taiko cannot handle, I need to provide the basic auth header when opening the page.)

### Versions
```
node --version
v10.15.3
taiko -v
Version: 1.0.0 (Chromium: 76.0.3803.0) RELEASE
```

**Comment by shubhamsc** [comment]
@F7502 When you open any URL with authorization token setting in headers, It'll replace if a token already exists with the same name. There are many ways to work with basic Http and bearer token together. See the recommended answers in this link https://stackoverflow.com/questions/22229996/basic-http-and-bearer-token-authentication. This may help you to resolve the issue.

**Comment by F7502** [comment]
@shubhamsc 
Thanks for the quick response. I investigated this some more and maybe I rephrase my issue a little bit.

How Chrome handles basic auth when I use it on my PC without taiko (i.e. me, clicking around with the mouse):
When I open the page for the first time, the server responds with `401 Unauthorized` and provides the `WWW-Authenticate` header. The browser opens a popup and asks me for username and password. I type it in and then the browser sends the initial request again, but now with the `Authorization: Basic ...` header containing my login credentials. For all subsequent requests (and this includes XHR requests), the browser automatically adds the same `Authorization` header from its local cache. But when I set the `Authorization` header explicitly in my Javascript code before sending a XHR request, this `Authorization` header is used instead of the cached one (this is the important part!).

How this works when using taiko:
I open the page calling `goto('https://example.com',{headers:{'Authorization': 'Basic xxx'}})`, so already the first request that is sent to the backend contains the correct auth credentials. For all subsequent requests, taiko also adds the same `Authorization` header to every request. If my Javascript code sends a XHR request and explicitly sets the `Authorization` header to some other value, **this is overwritten by taiko before sending the request.**

So in short, using `goto(...)` and providing an `Authorization` header with basic auth credentials does not result in the same behavior as providing credentials to the browser via the basic auth popup. The behavior of calling `goto(...)` with some predefined headers might be consistent or valid, but from my point of view it is not a feasible way of testing pages with basic auth (as recommended in #82).

Currently, `goto(...)` handles predefined headers as "always set before sending". What I need is "set before sending **if not already present**".

**Comment by negiDharmendra** [comment]
Right now Taiko uses [setExtraHTTPHeaders](https://github.com/getgauge/taiko/blob/4196c68d5adb36d3264b90a1b1f9735efe0610a7/lib/taiko.js#L881) API of [CDP's](https://chromedevtools.github.io/devtools-protocol) Network domain. Which forces all the subsequent requests to have the header configured by `setExtraHTTPHeaders` API.

**Example**

```
await openBrowser();
await goto('example.com, { headers: { Authorization: 'Basic <base64 encoded string>' } });
await goto('example-1.com');
```
In the above examples, we are setting `Authorization` header for `example.com` which forces all the requests after that to have the same `Authorization` header even if for some `ajax` request user has configured a different value of `Authorization` header.
Even all the requests for `example-1.com` will have the same header. This seems to be wrong because those are two different domains and the header configured for one domain should not be applied to another domain.


**Approaches to fix this issue.**

Taiko can keep a record of headers and the domain for which the header is configured. And apply all the domain specific header to every request if not already present.

**Comment by F7502** [comment]
Thx, looking forward to the next release :-)

**Comment by negiDharmendra** [comment]
@F7502 We have released a new [version](https://github.com/getgauge/taiko/releases/tag/v1.0.6) of Taiko. Please let us know if this new version resolves your issue.

**Comment by F7502** [comment]
@negiDharmendra 

> Please let us know if this new version resolves your issue.

Yes, works as expected now, many thanks for the fix :-)
